src/v2.rs

```diff
@@ -556,46 +692,70 @@ impl EszipV2 {
       bytes.extend_from_slice(string.as_bytes());
     }
 
-    // We serialize as ESZIP2.1 only if we have an npm snapshot. This is to
-    // maximize backwards compatibility of newly created archives with older
-    // deserializers.
-    let has_npm_snapshot = self.npm_snapshot.is_some();
-    let mut header = if has_npm_snapshot {
-      ESZIP_V2_1_MAGIC.to_vec()
-    } else {
-      ESZIP_V2_MAGIC.to_vec()
-    };
-    header.extend_from_slice(&[0u8; 4]); // add 4 bytes of space to put the header length in later
+    let (source_hash_function, source_digest_size) = self
+      .options
+      .source_hash_function
+      .zip(self.options.source_digest_size())
+      .expect("Disabling checksum should not be possible");
+
+    debug_assert_eq!(
+      source_digest_size,
+      source_hash_function.digest_size(),
+      "customizing the digest size should not be posible"
+    );
+
+    let mut options_header = ESZIP_V2_2_MAGIC.to_vec();
+
+    let options_header_length_pos = options_header.len();
+    const OPTIONS_HEADER_LENGTH_SIZE: usize = size_of::<u32>();
+    options_header.extend_from_slice(&[0; OPTIONS_HEADER_LENGTH_SIZE]); // Reserve for length
```

Comment:
Note to reviewer: right not length is known, but the nature of `Options` is thought of as variable length.